### DISCUSSION TITLE
Slicing image into 80x40=3200 parts, 1 px padding, custom naming the output

### DISCUSSION THREAD

**Discussion by JohanBurnetPayano** [ImageMagick/ImageMagick#7360] [Help]
### ImageMagick version

7.1.1-33 Windows Self Installer

### Operating system, version and so on

Windows 11 64 Bit

### Description

For a personal project I need to create tiles that will be loaded into a program. The program requires the tiles to be a particular size and also have 1 pixel of padding on all sides. The images also need to follow a naming convention so the program puts them in a correct order. The image in question is 81920x40960 and needs to be cut into 1024x1024 tiles which result in 80x40 tiles, a total of 3200 tiles. Would be a simple job if the 1 pixel padding wasn't a requirement, f you measured 3 tiles across on the map at 1:1 it would be 3066 pixels ((1024-2)*3). The naming convention also makes it tricky, it should look like this: "m000000" where the first three 0s is the x coordinate and the latter three the y coordinates. Manually renaming 3200 tiles is a job I would like to avoid.

I have heard that imagemagick is able to do something like this, so far... I got the following prompt except, it isn't working for me.

`convert {BaseImageFilePath} -crop 1024x1024+{1020*X}+{1020*Y}! -quality 90 {OutputDirectory}\{MapPrefix}{X}{Y}.jpg`

May a fancy person tell me what I am doing wrong? I've banged my head against this one for about a week now to no avail.

**Comment by snibgo** [comment]
Use `magick`, not `convert`.

>  "m000000" where the first three 0s is the x coordinate and the latter three the y coordinates.

I don't understand that. 3 digits is not enough to record numbers like 1024, 2048, etc.

> ... and also have 1 pixel of padding on all sides. 

I don't understand that requirement. What size do you want the output images to be? Do you want to cut the input into squares that are 1024x1024 pixels, and then add a border to each square image, before saving them?

Assuming you want to add a red border:
```
magick input.png -crop 1024x1024 -set filename:tile "%[fx:page.x]_%[fx:page.y]" +repage -bordercolor Red -border 1 myfile_%[filename:tile].png
```
If used in a Windows BAT script, double each percent: `%%`.

The output filenames will be like `myfile_2048_4096.png`, where the numbers are the offsets into the input image. There isn't a way within IM to add leading zeros to the numbers. You can use a drive and directory in the input and output filenames.

If you really must have leading zeros, you could do that in a shell loop. You could call `magick` and read the large input file 3200 times, which would take a long time. An alternative is to use the command I showed, then process each file in a script, renaming it. That would be much faster.

**Reply by JohanBurnetPayano** [reply]
Thanks for the reply, allow me the clarify your questions.

> I don't understand that. 3 digits is not enough to record numbers like 1024, 2048, etc.

The program requires the tiles to be named a certain way so they appear in a certain order. I.E: position x-16, y-3 would correspond to m016003 and m000000 would be the very first one.

> ... and also have 1 pixel of padding on all sides.

Every tile needs to have an empty or transparent 1 pixel border. Otherwise they will overlap over that one single pixel and over the course of 80 tiles that would mean 160 pixels.

I will try and use your prompt and see what I can get with that. Thanks again boss.

**Comment by GeeMack** [comment]
With ImageMagick you can crop an image into a grid of "X" units by "Y" units and name each output file according to where that unit originated. This command will crop the input into 80 units wide by 40 units high, then give each one a file name reflecting its original row and column position on the grid...  

    magick input.png -crop 80x40@ -set filename:0 %[fx:int(t/80)+100]x%[fx:(t%80)+100] image_%[filename:0].png


That reads the image and crops it into 80 by 40 units. (Caveat: If the width or height of the input image isn't evenly divisible by the number of units, there may be a single pixel difference in the dimensions of the pieces.)

Then it uses some FX formulas to create the output file names that will look like this...

    image_100x100.png
    image_100x101.png
    image_100x102.png
    ...
    image_139x177.png
    image_139x178.png
    image_139x179.png

Since ImageMagick can't really do much in the way of text formatting, I added 100 to each output file name calculation to make sure all the numbers have three digits.

I don't understand what you mean by padding, but each image could easily be shaved and/or bordered by a particular amount before writing the outputs. You'll have to be more specific with your question.

EDITED TO ADD:

This command will crop the input image into 80x40 tiles. Then inside parentheses it takes a clone of the input, rolls it -1 pixel up and -1 pixel left, and crop it into 80x40 tiles. Then it adds a single row of pixels to the top and left of each tile.

After the parentheses, with some "-reverse" operations to adjust the order, it composites the cropped input image tiles over those rolled and spliced tiles, leaving each tile with a single additional pixel on both the right and bottom that overlaps the adjacent image(s).

    magick input.png -write mpr:img0 -crop 80x40@ ^
       ( +clone mpr:img0 -roll -1-1 -crop 80x40@ -splice 1x1 null: +insert ) ^
       -reverse +repage -layers composite -reverse ^
       -set filename:0 %[fx:int(t/80)+100]x%[fx:(t%80)+100] ^
          image_%[filename:0].png

**Reply by JohanBurnetPayano** [reply]
Thank you for your reply,

The text formatting is important because it tells the image where it should appear. m000000.JPG would be the top left tile. m001000 would be one tile to the right.

> I don't understand what you mean by padding, but each image could easily be shaved and/or bordered by a particular amount before writing the outputs. You'll have to be more specific with your question.

Every "tile" needs 1 pixel on each side from the adjacent tile so they blend properly. So if you measured 3 tiles across on the map at 1:1 it would be 3066 pixels ((1024-2)*3). It makes cutting up a large image very tedious.

I will try to use this prompt and will report on the results.

**Reply by GeeMack** [reply]
Read my EDIT to see a method for adding that single pixel overlap to each tile.

**Reply by JohanBurnetPayano** [reply]
Thanks a lot for the help. I seem to be having hardware problems with RAM now. So I can not test your solution.

**Comment by fmw42** [comment]
By padding do you mean some constant color pad or overlap with the next section in x and y?  If the latter, how should the cells at the boundaries of the image be handled -- ignore overlap on that side or reflect the image to pad or pad with some constant color?

**Reply by JohanBurnetPayano** [reply]
I believe for the tiles to blend with each other there needs to be a transparent pixel between them. Otherwise the overlap will stack and since there are 80 in a row and 40 in a column that results in a lot of overlap.

**Comment by fmw42** [comment]
I do not know if this will help you, but I have a bash unix shell script called overlapcrop that might do what you want.  However, it only runs on Linux, Mac OSX, and Windows 10/11 Unix. It does not run under Windows DOS CMD.  If interested, see http://www.fmwconcepts.com/imagemagick/overlapcrop/index.php

**Reply by JohanBurnetPayano** [reply]
It might very well help. Worth a try, thanks for the offer boss!

**Comment by snibgo** [comment]
> Every "tile" needs 1 pixel on each side from the adjacent tile so they blend properly. 

> I believe for the tiles to blend with each other there needs to be a transparent pixel between them.

These seem to be contradictory. Adding a transparent pixel is easy.

An overlap, so pixels at the edge of each tile are duplicated at adjacent tiles, is more awkward. This Windows BAT script, `cropOverlap.bat`, will do that. It works by building an IM script, then running it. The large number of pixels in the input image are read only once.
```
rem Crops input image into overlapping pieces.

rem %1 input image
rem %2 prefix for output files, eg mydir\m
rem %3 width of each output piece
rem %4 height of each output piece
rem %5 horizontal overlap
rem %6 vertical overlap

echo off

setlocal enabledelayedexpansion

set INFILE=%1
set OUTPREF=%2
set W1=%3
set H1=%4
set HO=%5
set VO=%6

set ScrFile=cropOverlap.scr

del %ScrFile% 2>nul

for /F "usebackq" %%L in (`%IMG7%magick ^
  %INFILE% -ping ^
  -format "Win=%%w\nHin=%%h\n" ^
  info:`) do set %%L

for /F "usebackq" %%L in (`%IMG7%magick ^
  xc: ^
  -format "LastHor=%%[fx:ceil(%Win%/(%W1%-%HO%))-1]\nLastVert=%%[fx:ceil(%Hin%/(%H1%-%VO%))-1]\n" ^
  info:`) do set %%L

echo LastHor=%LastHor% LastVert=%LastVert%

(
  echo %INFILE% -write mpr:INP +delete
  for /L %%A in (0,1,%LastVert%) do (
    set /A OffsY=^(%H1%-%VO%^)*%%A
    set LZY=00%%A
    set LZY=!LZY:~-3!

    for /L %%B in (0,1,%LastHor%) do (
      set /A OffsX=^(%W1%-%HO%^)*%%B
      set LZX=00%%B
      set LZX=!LZX:~-3!
      echo ^( mpr:INP -crop %W1%x%H1%+!OffsX!+!OffsY! +repage -write %OUTPREF%!LZX!!LZY!.png +delete ^)
    )
  )
  echo -exit
) >>%ScrFile%

type %ScrFile%

%IMG7%magick -script %ScrFile%

if ERRORLEVEL 1 (
  echo %0: Problem in magick command
  exit /B 1
)

echo on

@endlocal
```

**Reply by JohanBurnetPayano** [reply]
Thanks a lot for the help. I seem to be having hardware problems with RAM now. So I can not test your solution.
